### ISSUE TITLE
Create matchers for `vitest`

### ISSUE THREAD

**Issue by ThomasAribart** [m-radzikowski/aws-sdk-client-mock#139]
Hello and thanks for this great lib 🙌

On my project we went from `jest` to `vitest`, mainly for performances. Sadly, we lost the `toHaveReceivedCommand` matchers in the process 🥲

Would be awesome to have them for `vitest` as well! Maybe it was the aim of separating `jest` matchers in another lib in the first place ?

See https://vitest.dev/guide/extending-matchers.html

Cheers !

**Comment by kldavis4** [comment]
I found this in the vitest docs: https://vitest.dev/guide/features.html#chai-and-jest-expect-compatibility

When I set `test.globals: true` in `vite.config.js`, the `aws-sdk-client-mock-jest` matchers seem to work.

**Comment by m-radzikowski** [comment]
Hey, sorry for the late reply. Thank you @kldavis4 for finding and sharing the way to make `aws-sdk-client-mock-jest` work with `vitest`. I will add that to the README.

While I see a raising popularity of `vitest`, I'm not using it right now myself. If anyone wants to add `vitest` matchers, or, even better, adjust current matchers so they can be use in `vitest` natively (if possible) - feel free to open a PR.

**Comment by colesiegel** [comment]
It still doesn't work for me, even with test.globals set to true.


`expect(ddbMock).toHaveReceivedCommandWith(QueryCommand, {});`

`Property 'toHaveReceivedCommandWith' does not exist on type 'Assertion<AwsStub<ServiceInputTypes, ServiceOutputTypes>>'.ts(2339)`

```
import { defineConfig } from 'vitest/config';

export default defineConfig({
  test: {
    reporters: ['junit', 'default'],
    outputFile: './.vitest/results.xml',
    coverage: {
      // include coverage even if no tests exist
      all: true,
      include: ['src/*'],
      reportsDirectory: './.vitest/coverage',
    },
    setupFiles: './src/setupTests.ts',
    // [https://github.com/m-radzikowski/aws-sdk-client-mock/issues/139]
    globals: true,
  },
});
```

**Comment by mbtts** [comment]
This is really excellent test utility, thank you to all who have contributed so far.

On this point from the README:

https://github.com/m-radzikowski/aws-sdk-client-mock/blob/c6896a392987cd3eefcfeea800bd9599464bb6fc/README.md?plain=1#L513-L514

One problem here is the [vitest “jest” matcher utils](https://github.com/vitest-dev/vitest/blob/ba1e7335f93b82c110209a6b6e3099908aa1b2f8/packages/expect/src/jest-matcher-utils.ts) are a much smaller subset of the [jest matcher utils](https://github.com/jestjs/jest/blob/be7e797bd6d1ef3476f1d9a15fb50fd03527bd19/packages/jest-matcher-utils/src/index.ts#L318) (which also rely on other Jest packages like [jest-diff](https://www.npmjs.com/package/jest-diff)).

✅ If the expectation passes there is no problem, the matchers work as expected with vitest.

❌ However if the expectations then [`printDiffOrStringify`](https://github.com/jestjs/jest/blob/be7e797bd6d1ef3476f1d9a15fb50fd03527bd19/packages/jest-matcher-utils/src/index.ts#L318) is called:

https://github.com/m-radzikowski/aws-sdk-client-mock/blob/c6896a392987cd3eefcfeea800bd9599464bb6fc/packages/aws-sdk-client-mock-jest/src/jestMatchers.ts#L426

At this point Vitest runner will bail:

```
TypeError: this.utils.printDiffOrStringify is not a function
```

---

Happy to stick my hand to contribute.

My thoughts are there are a few possibilities:

1. Avoid: Feature-sniff and use an alternative function/output when used in a non-Jest environment/Jest-like environment
2. Import: [jest-utils](https://www.npmjs.com/package/jest-utils) could become a dependency (and its implementation of [`printDiffOrStringify`](https://github.com/jestjs/jest/blob/be7e797bd6d1ef3476f1d9a15fb50fd03527bd19/packages/jest-matcher-utils/src/index.ts#L318) used verbatim)
3. Delegate: Approach the Vitest maintainers about providing the same functionality/API (I could not see anything planned/open issue, but might have overlooked something)

In the very short term I think either the limitations should be documented _-or-_ vitest support not officially documented as supported for just the matchers.

---

@colesiegel On the typings question, I hope you got this resolved? Sharing my configuration in case it helps:

* `@aws-sdk/`: 3.409.0
* `aws-sdk-client-mock-jest`: 3.0.0
* `aws-sdk-client-mock`: 3.0.0
* `typescript`: 5.2.2
* `vitest`: 0.34.3

**vitest.config.ts**:
```ts
import { defineConfig } from 'vitest/config';

export default defineConfig({
  test: {
    globals: true,
    mockReset: true,
    setupFiles: ['./awsSdkClientMatchers.ts'],
  },
});
```

**awsSdkClientMatchers.ts**:
```ts
import 'aws-sdk-client-mock-jest';
```

**tsconfig.json**:
```json
{
  "extends": "@tsconfig/node18/tsconfig",
  "compilerOptions": {
    "allowJs": true,
    "lib": ["ESNext"],
    "module": "NodeNext",
    "moduleResolution": "nodenext",
    "noEmit": true,
    "types": ["vitest/globals"]
  }
}
```

**Comment by douglasg14b** [comment]
@m-radzikowski Can this be re-opened? Vitest matchers don't yet exist for this project, and only kind-of work as a side effect of Vitest being very diligent about being Jest-compatable.

Dedicated Vitest matchers would go a long ways, keeping the issue open may help with fishing for a contribution 🤔

**Comment by EHadoux** [comment]
> It still doesn't work for me, even with test.globals set to true.
> 
> `expect(ddbMock).toHaveReceivedCommandWith(QueryCommand, {});`
> 
> `Property 'toHaveReceivedCommandWith' does not exist on type 'Assertion<AwsStub<ServiceInputTypes, ServiceOutputTypes>>'.ts(2339)`

I still see this with 4.1.0. It works when running the tests but TS is not happy in vscode. Any chance someone knows how to fix it?

Using 
```typescript
import { DynamoDBDocumentClient, UpdateCommand } from '@aws-sdk/lib-dynamodb';
import { mockClient } from 'aws-sdk-client-mock';
import 'aws-sdk-client-mock-jest/vitest';
import { beforeEach, describe, expect, it } from 'vitest';

describe('something', () => {
  const ddbMock = mockClient(DynamoDBDocumentClient);

  beforeEach(() => {
    ddbMock.reset();
  });

  it('something', async () => {
    expect(ddbMock).toHaveReceivedCommandWith(UpdateCommand, {});
  });
});
```

**Comment by ColinMorris83** [comment]
Yes I am getting the same, the tests will run ok, but show errors in VSCode on the expect lines.
A workaround at the moment seems to be to import both
```
import 'aws-sdk-client-mock-jest';
import 'aws-sdk-client-mock-jest/vitest';
```

**Comment by piersroberts** [comment]
You can also fix TS issues by declaring the types in a vitest.d.ts file if you don't want to keep your test tidier

```
import 'vitest';
import { CustomMatcher } from 'aws-sdk-client-mock-jest';

declare module 'vitest' {
  type Assertion<T = unknown> = CustomMatcher<T>;
  type AsymmetricMatchersContaining = CustomMatcher;
}
```
